xarray/core/common.py

```diff
@@ -818,3 +819,34 @@ def ones_like(other, dtype=None):
     """Shorthand for full_like(other, 1, dtype)
     """
     return full_like(other, 1, dtype)
+
+
+def is_np_datetime_like(dtype):
+    """Check if a dtype is a subclass of the numpy datetime types
+    """
+    return (np.issubdtype(dtype, np.datetime64) or
+            np.issubdtype(dtype, np.timedelta64))
+
+
+def contains_netcdftime_datetimes(var):
+    """Check if a variable contains netcdftime datetime objects"""
+    from ..coding.times import _import_netcdftime_datetime
+    
+    try:
+        netcdftime_datetime = _import_netcdftime_datetime()
+    except ImportError:
+        return False
+    else:
+        return isinstance(var.data.flatten()[0], netcdftime_datetime)
+
+
+def contains_datetime_datetimes(var):
+    """Check if a variable contains datetime.datetime objects"""
+    return isinstance(var.data.flatten()[0], datetime)
+    
+
+def _contains_datetime_like_objects(var):
+    """Check if a variable contains datetime like objects (either
+    np.datetime64, np.timedelta64, or netcdftime._netcdftime.datetime)"""
+    return (is_np_datetime_like(var.dtype) or
+            contains_netcdftime_datetimes(var) or contains_datetime_datetimes(var))
```

Comment:
Would it be reasonable require using netcdftime datetimes, and remove the `contains_datetime_datetimes` check?

Reply:
Yes, I think that's a good idea.  I suspect the most common use cases for a `NetCDFTimeIndex` will be:
- No leap calendars
- Standard and proleptic Gregorian calendars with dates before 1678 or after 2262.

The first case will work out of the box here; the second case will not it all circumstances, because `netcdftime.num2date` returns `datetime.datetime` objects in some instances.  

Hopefully in newer versions of netcdftime (or `cftime` as it will likely be renamed to) we can add an option for `num2date` to only return `netcdftime.datetime` objects.  I have a PR open upstream, https://github.com/Unidata/netcdftime/pull/39, which implements just that.  If a version of that gets merged, then with the latest cftime library we'll be able to support both of the use cases above cleanly.

Tomorrow I'll remove the `contains_datetime_datetimes` logic, which will simplify the logic some and the roundtrip tests a fair amount.

Reply:
Do we need that PR before we can guarantee that a NetCDFTimeIndex always contains netcdftime.datetime objects?

Reply:
I think we can move forward without it (I'll make sure the logic is such that if `num2date` returns `datetime.datetime` objects we'll just use a generic `pd.Index`).  

Round tripping for dates of type `netcdftime.DatetimeGregorian` or `netcdftime.DatetimeProlepticGregorian` might need to wait for that PR, however.  Otherwise the logic could be quite messy, e.g. https://github.com/pydata/xarray/pull/1252#issuecomment-374711420.

Reply:
Right, I remember reading your earlier comment now.

One clean option would be to require the new `cftime` module (with your change) for this functionality. That way we can rely on standard behavior here.

Roundtripping is a very nice property to maintain when possible (if only for our own sanity!)

Reply:
> One clean option would be to require the new cftime module (with your change) for this functionality. That way we can rely on standard behavior here.

Yes, I think that's completely reasonable.

> Roundtripping is a very nice property to maintain when possible (if only for our own sanity!)

Totally agree!